api/src/main/java/org/dicekeys/api/DerivationRecipe.kt

```diff
@@ -1,52 +1,42 @@
-import kotlinx.serialization.json.int
-import kotlinx.serialization.json.jsonObject
-import kotlinx.serialization.json.jsonPrimitive
+import kotlinx.serialization.json.*
 import org.dicekeys.crypto.seeded.DerivationOptions
 
-fun addFieldToEndOfJsonObjectString(originalJsonObjectString: String, fieldName: String, fieldValue: String): String {
-    val lastClosingBraceIndex = originalJsonObjectString.lastIndexOf("}")
-    if (lastClosingBraceIndex < 0) return originalJsonObjectString
-    val prefixUpToFinalClosingBrace = originalJsonObjectString.substring(0, lastClosingBraceIndex)
-    val suffixIncludingFinalCloseBrace = originalJsonObjectString.substring(lastClosingBraceIndex)
-    val commaIfObjectNonEmpty = if (prefixUpToFinalClosingBrace.contains (':')) "," else  ""
-    return "$prefixUpToFinalClosingBrace$commaIfObjectNonEmpty\"$fieldName\":$fieldValue$suffixIncludingFinalCloseBrace"
+fun JsonObject.update(updateJsonObject: JsonObject): JsonObject {
+    return JsonObject(toMutableMap().apply {
+        updateJsonObject.forEach { (key, value) ->
+            put(key, value)
+        }
+    })
 }
 
-fun addLengthInCharsToDerivationOptionsJson(derivationOptionsWithoutLengthInChars: String, lengthInChars: Int = 0 ): String {
-    if (lengthInChars <= 0) return derivationOptionsWithoutLengthInChars
-    return addFieldToEndOfJsonObjectString(derivationOptionsWithoutLengthInChars, "lengthInChars", lengthInChars.toString())
+fun JsonObject.canonicalize(): String{
+    return toCanonicalizeRecipeJson(this)
 }
 
-fun addLengthInBytesToDerivationOptionsJson(derivationOptionsWithoutLengthInBytes: String, lengthInBytes: Int = 0 ): String {
-    if (lengthInBytes <= 0) return derivationOptionsWithoutLengthInBytes
-    return addFieldToEndOfJsonObjectString(derivationOptionsWithoutLengthInBytes, "lengthInBytes", lengthInBytes.toString())
+fun JsonObjectBuilder.addLengthInCharsToDerivationOptionsJson(lengthInChars: Int){
+    if(lengthInChars > 0){
+        put("lengthInChars", lengthInChars)
+    }
 }
 
-fun addSequenceNumberToDerivationOptionsJson(derivationOptionsWithoutSequenceNumber: String, sequenceNumber: Int): String {
-    if (sequenceNumber == 1) return derivationOptionsWithoutSequenceNumber
-    return addFieldToEndOfJsonObjectString(derivationOptionsWithoutSequenceNumber, "#", sequenceNumber.toString())
+fun JsonObjectBuilder.addLengthInBytesToDerivationOptionsJson(lengthInBytes: Int){
+    if(lengthInBytes > 0){
+        put("lengthInBytes", lengthInBytes)
+    }
 }
 
-private fun augmentRecipeJson(template: DerivationRecipe, sequenceNumber: Int, lengthInChars: Int, lengthInBytes: Int): String {
-    var recipeJson = template.recipeJson
-    if (template.type == DerivationOptions.Type.Password && lengthInChars > 0) {
-        recipeJson = addLengthInCharsToDerivationOptionsJson(recipeJson, lengthInChars)
+fun JsonObjectBuilder.addSequenceNumberToDerivationOptionsJson(sequenceNumber: Int){
+    if(sequenceNumber > 1){
```

Comment:
If the object has a sequenceNumber of 2, and this is called with sequenceNumber of 1, will this remove the old sequenceNumber as it should?

Reply:
Good catch. Changed the implementation and added unit tests.